pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -326,11 +327,54 @@ func (c *Cluster) postReconcileUpdateOSDProperties(desiredOSDs map[int]*OSDInfo)
 	}
 	logger.Debugf("post processing osd properties with %d actual osds from ceph osd df and %d existing osds found during reconcile", len(osdUsage.OSDNodes), len(desiredOSDs))
 	for _, actualOSD := range osdUsage.OSDNodes {
-		if desiredOSD, ok := desiredOSDs[actualOSD.ID]; ok {
-			if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
-				// Log the error and allow other updates to continue
-				logger.Error(err)
-			}
+		if err := c.resizeOsdCrushWeight(actualOSD); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Error("failed to resize osd crush weight on cluster in namespace %s %v", c.clusterInfo.Namespace, err)
+		}
+		desiredOSD, ok := desiredOSDs[actualOSD.ID]
+		if !ok {
+			continue
+		}
+		if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Error("failed to update device class on cluster in namespace %s %v", c.clusterInfo.Namespace, err)
+		}
+	}
+	return nil
+}
+
+func convertKibbytesToTibbytes(kib string) (float64, error) {
+	kibFloat, err := strconv.ParseFloat(kib, 64)
+	if err != nil {
+		return float64(0), fmt.Errorf("failed converting string to float: %v", err)
+	}
+	return kibFloat / float64(units.GiB), nil
+}
+
+func (c *Cluster) resizeOsdCrushWeight(actualOSD cephclient.OSDNodeUsage) error {
+	if !c.spec.Storage.AllowOsdCrushWeightUpdate {
+		return nil
+	}
+	if actualOSD.KB != actualOSD.CrushWeight {
+		desiredOSDTiB, err := convertKibbytesToTibbytes(actualOSD.KB.String())
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert kib to tib")
+		}
+		desiredOSDTiB = desiredOSDTiB / float64(units.GiB)
+		crushWeight, err := strconv.ParseFloat(actualOSD.CrushWeight.String(), 64)
+		if err != nil {
+			return fmt.Errorf("failed converting string to float: %v", err)
+		}
+		// do not reweight if the desired crush weight is 0 or less than the current crush weight
+		if desiredOSDTiB == float64(0) || desiredOSDTiB < crushWeight {
+			return nil
```

Comment:
Let's log if the value is not being set, so we can see why it was skipped if it was non-zero

Reply:
Actually how about if we make this an error condition if `desiredOSDTiB < crushWeight`? Then we will anyway log the issue, but not fail the reconcile. Also, maybe move this check inside `convertKibbytesToTibbytes()` so you can easily unit test this condition.

Reply:
Added the error log, but lets keep it outside for better function readability